aggregator/src/aggregator.rs

```diff
@@ -2214,6 +2243,40 @@ impl VdafOps {
                                 )
                             )?;
 
+                            // To ensure that concurrent aggregations don't write into a
+                            // currently-nonexistent batch aggregation, we write (empty) batch
+                            // aggregations for any that have not already been written to storage.
```

Comment:
I think I get why this is important for the time interval case. Suppose I have contiguous batch aggregations:
```
    b1         b2       b3
|---------|---------|---------|
t0       t2         t3        t4
```

And suppose there are some reports in `b1` and `b3` but none in `b2`. If we service a collection over `[t0, t4)`, then `batch_aggregations` will be `[b1, b2, b3]`, and we'll construct a collection over the reports in `b1` and `b3`. We write out an empty, collected batch aggregation for `b2` to ensure we later will refuse to collect the interval `[t2, t3)`.

But am I right in thinking that for the fixed size query case, this shouldn't be necessary? In that case `Q::get_batch_aggregations_for_collect_identifier` will always yield a single batch aggregation, right?

I'm not proposing any changes here, just want to make sure I understand what this does.

Reply:
I don't think this is so much for future collections over `[t2, t3)` as it is for concurrent aggregations landing in that interval. By responding to an absence of a row with insertion of a "zero" row, we are making phantom read anomalies irrelevant, by forcing the issue and writing relevant rows. If there's a concurrent aggregation at the same time as this larger collection, then either the collection wins, and the zero batch aggregation gets written, filling in the hole, and the aggregation retries, then gets a batch collected error, or the aggregation wins, writing a non-zero batch aggregation into the hole, so the collection retries, reads the now-finished aggregation in its new snapshot, and succeeds.

Without this write, phantom read anomalies mean that we can't be sure from a collection transaction that we see all report aggregations and batch aggregations, thus it's possible we are seeing a unique view of a batch that is missing certain reports compared to future transactions, breaking our replay protections.

Reply:
You have the right idea as to what batch aggregations end up written; David has the right idea as to why. We need an explicit row to exist in order to trigger a write-conflict failure; writing "empty" batch aggregations provides these rows.

This could still impact fixed size in a few ways:
* Batch aggregation rows for a given batch are sharded; through bad luck or low report volume, some batch aggregation shards might end up with no reports assigned to them, and would therefore not be written by the time this code runs. The empty-batch-aggregations functionality would fill in rows for any such shards.
* A Collector could query for a batch that has no reports, as is done in your initial example (`[t2, t3]` is a batch with no reports). This affects both query types, though I think triggering this behavior for fixed-size queries would require some extra-DAP backchannel to communicate the relevant batch ID to the collector, and even then I suppose we'd need to have `max_batch_size = 0` (...and to fix up our code to support setting `max_batch_size = 0`...) for the collection to actually succeed. I think this one is mostly hypothetical, at the moment, for fixed-size.

Reply:
Ok, thanks for the explanations. At the risk of introducing comments that might get out of sync with the code, should all this be documented, since deleting or modifying this code could cause subtle problems that won't manifest obviously in tests?

Reply:
We now have tests that check the expected interaction between aggregation & collect jobs. What documentation would you like to see? The remainder of the behavior is standard Postgres behavior, and we document in code [why we trigger retries](https://github.com/divviup/janus/blob/3d15c87f7051acca0416c7948410b5076c958cc7/aggregator/src/aggregator/accumulator.rs#L118-L140) in some cases. We might also trigger retries if two writers attempt to write into the same already-existing batch aggregation row, but that's a standard serialization failure and will be true for all interactions with our DB.

Reply:
Well, what I'm worried about is someone in the future deleting or re-ordering this code, or moving it to another transaction because they don't realize the importance of _where_ it happens. So perhaps a paragraph explaining that these empty batch aggregations need to be written out in this particular transaction?